js/iframe-coordinator/libs/workers/worker-manager.ts

```diff
@@ -0,0 +1,288 @@
+} from './constants';
+import * as SpawnWorker from './spawn-worker.worker.ts';
+
+export const WORKER_MESSAGE_EVENT_TYPE: string = 'workermessage';
+
+const DEFAULT_ERROR_WINDOW_COUNT_THRESHOLD = 10;
+const DEFAULT_ERROR_WINDOW_MILLIS = 30000;
+
+/*
+ * TODO for this feature:
+
+ * get bi-directional comms working
+ * Finalize backgroundClient api
+   * Check on impl interface idea and add NavigationRequester interface
+   * should navRequest be url based?
+   * Will docs get pulled in correctly? pull them from client.ts
+ * Message validation
+  * Add checks to ensure lifecycle is not updated to prior state?
+  * Ensure payload is valid for message types
+  * Better checks to ensure all enums and cases are handled?
+  * stronger types of messages between this manager and the workers?
+ * docs
+ *   dont forget that this is sorta dangerous with indexeddb
+ *
+ * Is the import of the worker good enough?
+  * Decide on class or text of the worker
+ * Make sure es6 target is ok
+ * Make sure it works in IE
+ *  check blob and fallback
+ * Improve worker demo for the real, useful demo.
+ *  use subscribe?
+ *  allow dynamic addition?
+ * Tests
+ *  tests for error rate
+ * better logging (these are important)
+ * support lenient shutdown?
+ *  notification on shutdown?
+ *  soft stop
+ *  allow restart?
+ * support client stop request?
+ *   worker never can close now as i have attached listeners
+ *   but, you can't re-add now either via registerclients
+ */
+
+enum WorkerPhase {
+  LOADING,
+  LOADED,
+  BOOTSTRAPPING,
+  RUNNING
+  // TODO Might want to implement this.  give it a chance to shutdown.  Need a timeout
+  // STOPPING
+}
+
+interface ManagedWorker {
+  id: string;
+  url: string;
+  phase: WorkerPhase;
+  worker: Worker;
+  errorWindowCount: number;
+  errorWindowStartTimestamp: number;
+}
+
+export default class WorkerManager implements EventListenerObject, EventTarget {
+  private static trackedWorkerEventTypes = ['message', 'error'];
+  private static evtTargetDelegate: DocumentFragment = null;
+  private static ID_PREFIX = 'iframeCoordinatorWorker-';
+  private static _workerIndex = 0;
+  private _workers: ManagedWorker[];
+  private _errorWindowCountThreshold = 0;
+  private _errorWindowMillis = 0;
+
+  constructor(
+    errorWindowCountThreshold: number = DEFAULT_ERROR_WINDOW_COUNT_THRESHOLD,
+    errorWindowMillis: number = DEFAULT_ERROR_WINDOW_MILLIS
+  ) {
+    this._errorWindowCountThreshold = errorWindowCountThreshold;
+    this._errorWindowMillis = errorWindowMillis;
+
+    if (WorkerManager.evtTargetDelegate === null) {
+      WorkerManager.evtTargetDelegate = document.createDocumentFragment();
+    }
+
+    this._workers = [];
+  }
+
+  public start(url: string): string {
+    const id = `${WorkerManager.ID_PREFIX}${++WorkerManager._workerIndex}`;
+
+    // @ts-ignore
+    const worker = new SpawnWorker();
+    WorkerManager.trackedWorkerEventTypes.forEach(curr => {
+      worker.addEventListener(curr, this);
```

Comment:
I call this out just because I thought it was cool and I didn't previously know you could attach an object implementing `EventListenerObject`.

This is cool because it makes it trivial to remove listeners for clean-up.  No need to track function references!